kani-compiler/src/codegen_boogie/context/boogie_ctx.rs

```diff
@@ -13,14 +29,320 @@ pub struct BoogieCtx<'tcx> {
+    }
+
+    pub fn codegen_rvalue(&self, rvalue: &Rvalue<'tcx>) -> Expr {
+        debug!(rvalue=?rvalue, "codegen_rvalue");
+        match rvalue {
+            Rvalue::Use(operand) => self.codegen_operand(operand),
+            Rvalue::BinaryOp(binop, box (lhs, rhs)) => self.codegen_binary_op(binop, lhs, rhs),
+            _ => todo!(),
+        }
+    }
+
+    pub fn codegen_binary_op(
+        &self,
+        binop: &BinOp,
+        lhs: &Operand<'tcx>,
+        rhs: &Operand<'tcx>,
+    ) -> Expr {
+        match binop {
+            BinOp::Eq => Expr::BinaryOp {
+                op: BinaryOp::Eq,
+                left: Box::new(self.codegen_operand(lhs)),
+                right: Box::new(self.codegen_operand(rhs)),
+            },
+            _ => todo!(),
+        }
+    }
+
+    pub fn codegen_terminator(
+        &self,
+        local_decls: &LocalDecls<'tcx>,
+        term: &Terminator<'tcx>,
+    ) -> Stmt {
+        let _trace_span = debug_span!("CodegenTerminator", statement = ?term.kind).entered();
+        debug!("handling terminator {:?}", term);
+        match &term.kind {
+            TerminatorKind::Call { func, args, destination, target, .. } => self.codegen_funcall(
+                local_decls,
+                func,
+                args,
+                destination,
+                target,
+                term.source_info.span,
+            ),
+            TerminatorKind::Return => Stmt::Return,
+            _ => todo!(),
+        }
+    }
+
+    pub fn codegen_funcall(
+        &self,
+        local_decls: &LocalDecls<'tcx>,
+        func: &Operand<'tcx>,
+        args: &[Operand<'tcx>],
+        destination: &Place<'tcx>,
+        target: &Option<BasicBlock>,
+        span: Span,
+    ) -> Stmt {
+        debug!(?func, ?args, ?destination, ?span, "codegen_funcall");
+        let fargs = self.codegen_funcall_args(local_decls, args);
+        let funct = self.operand_ty(local_decls, func);
+        // TODO: Only hooks are handled currently
+        match &funct.kind() {
+            ty::FnDef(defid, substs) => {
+                let instance =
+                    Instance::expect_resolve(self.tcx, ty::ParamEnv::reveal_all(), *defid, substs);
+
+                if let Some(hk) = self.hooks.hook_applies(self.tcx, instance) {
+                    return hk.handle(self, instance, fargs, *destination, *target, Some(span));
+                }
+                todo!()
+            }
+            _ => todo!(),
+        }
+    }
+
+    pub fn codegen_funcall_args(
+        &self,
+        local_decls: &LocalDecls<'tcx>,
+        args: &[Operand<'tcx>],
+    ) -> Vec<Expr> {
+        debug!(?args, "codegen_funcall_args");
+        args.iter()
+            .filter_map(|o| {
+                let ty = self.operand_ty(local_decls, o);
+                // TODO: handle non-primitive types
+                if ty.is_primitive() {
+                    return Some(self.codegen_operand(o));
+                }
+                None
```

Comment:
maybe add a `todo!()`

Reply:
Unfortunately, I cannot add a `todo!` here because it'll break the few tests that currently work. This is because they all involve `kani::assert` which has a string slice argument that will result in a crash. I added a `TODO` comment though.